addons/binding/org.openhab.binding.osramlightify/src/main/java/org/openhab/binding/osramlightify/internal/LightifyConnector.java

```diff
@@ -0,0 +1,478 @@
+
+import org.openhab.binding.osramlightify.internal.LightifyTransmitQueue;
+import org.openhab.binding.osramlightify.internal.LightifyTransmitQueueSender;
+
+/**
+ * @author Mike Jagdis - Initial contribution
+ */
+public final class LightifyConnector extends Thread implements LightifyTransmitQueueSender<LightifyMessage> {
+
+    /**
+     * The port the gateway listens on.
+     */
+    private static final int LIGHTIFY_GATEWAY_PORT = 4000;
+
+    /**
+     * The maximum time we wait for data from the gateway before breaking off to
+     * check for interrupts or to see if it is time to poll for state updates.
+     *
+     * The smaller this is the more cpu we will use but the faster we will respond
+     * to requests to stop the connector thread (for instance when reloading the
+     * binding). It MUST be less than SafeMethodCaller.DEFAULT_TIMEOUT in the core.
+     */
+    private static final int MAX_PRE_MESSAGE_TIMEOUT = 4800; // milliseconds
+    /**
+     * How long to wait for data while in the middle of reading a message
+     * from the gateway.
+     *
+     * This should never normally be reached. If it does it means the gateway
+     * has disappeared without notice in the middle of talking to us.
+     */
+    private static final int IN_MESSAGE_TIMEOUT = 5000; // milliseconds
+
+    /**
+     * How long to pause between reconnect attempts if the gateway does not seem
+     * to be responding.
+     */
+    private static final int RECONNECT_INTERVAL = 10000; // milliseconds
+
+    /**
+     * How many times to attempt reconnection before marking the gateway (and all
+     * paired devices) offline.
+     *
+     * This can be expensive and we _do_ queue commands so making this long
+     * enough to allow a gateway reboot makes sense.
+     */
+    private static final int RECONNECT_ATTEMPTS_BEFORE_OFFLINE = 3;
+
+    private static final int MAXIMUM_RESENDS = 5;
+
+    private final Logger logger = LoggerFactory.getLogger(LightifyConnector.class);
+
+    private LightifyBridgeHandler bridgeHandler;
+
+    private LightifyTransmitQueue<LightifyMessage> transmitQueue = new LightifyTransmitQueue<>(this);
+
+    private int reconnectAttempts = 1;
+
+    private boolean interrupted = false;
+
+    private Socket socket;
+    private InputStream in;
+    private OutputStream out;
+    private LightifyListPairedDevicesMessage initialPoll;
+
+    private final LightifyListGroupsMessage listGroups = new LightifyListGroupsMessage();
+    private final LightifyListPairedDevicesMessage listPairedDevices = new LightifyListPairedDevicesMessage();
+    private boolean havePoll = false;
+    private long nextPoll;
+    private long pollInterval;
+
+    private int resendCount = 0;
+    private int seqNo = 0;
+
+    private class ExceptionHandler implements Thread.UncaughtExceptionHandler {
+        public void uncaughtException(Thread thread, Throwable throwable) {
+            logger.error("Connector died: ", throwable);
```

Comment:
That's exactly the point of an uncaught exception handler! We should never get here. If we do it's because we are dying. Someone is going to want to know why. Not having an uncaught exception handler on a thread implies excessive confidence in potentially large amounts of code. It's the exception that proves the rule :-)

Reply:
I agree, but nevertheless you should use our scheduler and schedule a job. The scheduler will also take care of `Exception`s and log them with error level.

I just wanted to remind you that we do not want the `Exception` class to be caught and rather catch the specific ones.